# Mengukur Jarak

Mengukuran jarak data (Similarity dan Dissimilarity) adalah proses untuk menentukan seberapa mirip atau berbeda dua data atau lebih. Jarak ini dapat digunakan dalam berbagai algoritma pembelajaran mesin, seperti clustering, klasifikasi, dan regresi. Dalam mengukur jarak data terdapat beberapa tipe atribut seperti:

- Numerik
- Binary
- Categorical
- Ordinal

## Mengukur Jarak Data (Tipe Numerik)

Salah satu tantangan dalam era ini adalah database yang memiliki banyak tipe data. Mengukur jarak adalah komponen utama dalam algoritma *clustering* berbasis jarak. Algoritma *partitioning* seperti **K-Means**, **K-Medoids**, **Fuzzy C-Means**, dan **Rough Clustering** sangat bergantung pada perhitungan jarak untuk melakukan pengelompokkan data.

Misalkan terdapat dua buah vektor $v_1$ dan $v_2$ dalam ruang dimensi $n$:
$v_1 = (x_1, x_2, \dots, x_n)$
$v_2 = (y_1, y_2, \dots, y_n)$

Berikut adalah beberapa ukuran jarak untuk data tipe numerik:

### 1. Minkowski Distance
Merupakan generalisasi dari jarak Euclidean dan Manhattan.
$$d_{min} = \left( \sum_{i=1}^{n} |x_i - y_i|^m \right)^{1/m}, m \geq 1$$

### 2. Manhattan Distance
Merupakan kasus khusus dari Minkowski Distance pada $m = 1$. Jarak ini sensitif terhadap *outlier*.
$$d_{man} = \sum_{i=1}^{n} |x_i - y_i|$$

### 3. Euclidean Distance
Jarak yang paling terkenal untuk data numerik (kasus khusus Minkowski pada $m = 2$). Sangat umum digunakan namun memiliki kelemahan jika fitur memiliki skala yang berbeda jauh (perlu normalisasi).
$$d_{euc} = \sqrt{\sum_{i=1}^{n} (x_i - y_i)^2}$$

### 4. Average Distance
Versi modifikasi dari Euclidean untuk memperbaiki hasil dengan menghitung rata-rata.
$$d_{ave} = \left( \frac{1}{n} \sum_{i=1}^{n} (x_i - y_i)^2 \right)^{1/2}$$

### 5. Weighted Euclidean Distance
Digunakan jika setiap atribut memiliki tingkat kepentingan (bobot) yang berbeda.
$$d_{we} = \left( \sum_{i=1}^{n} w_i (x_i - y_i)^2 \right)^{1/2}$$

### 6. Chord Distance
Modifikasi Euclidean untuk mengatasi kekurangan pada data yang tidak dinormalisasi.
$$d_{chord} = \left( 2 - 2 \frac{\sum_{i=1}^{n} x_i y_i}{\|x\|_2 \|y\|_2} \right)^{1/2}$$

### 7. Mahalanobis Distance
Mengukur jarak yang memperhitungkan korelasi antar fitur menggunakan matriks kovarians ($S$).
$$d_{mah} = \sqrt{(x - y) S^{-1} (x - y)^T}$$

### 8. Cosine Similarity
Banyak digunakan untuk kemiripan dokumen (teks).
$$Cosine(x, y) = \frac{\sum_{i=1}^{n} x_i y_i}{\|x\|_2 \|y\|_2}$$

### 9. Pearson Correlation
Menghitung kemiripan antara dua pola (sering digunakan pada ekspresi gen).
$$Pearson(x, y) = \frac{\sum_{i=1}^{n} (x_i - \mu_x)(y_i - \mu_y)}{\sqrt{\sum_{i=1}^{n} (x_i - \mu_x)^2} \sqrt{\sum_{i=1}^{n} (y_i - \mu_y)^2}}$$

## Mengukur Jarak Atribut Binary
Atribut biner hanya memiliki dua status (0 dan 1). Perhitungannya menggunakan tabel kontingensi $2 \times 2$.
Misalkan:
- $q$: jumlah atribut yang bernilai 1 untuk kedua objek.
- $r$: jumlah atribut yang bernilai 1 untuk objek $i$ tapi 0 untuk objek $j$.
- $s$: jumlah atribut yang bernilai 0 untuk objek $i$ tapi 1 untuk objek $j$.
- $t$: jumlah atribut yang bernilai 0 untuk kedua objek.

### 1. Simetris Biner (Symmetric Binary)
Digunakan jika nilai 0 dan 1 dianggap sama pentingnya (misal: jenis kelamin).
$$d(i, j) = \frac{r + s}{q + r + s + t}$$

### 2. Asimetris Biner (Asymmetric Binary)
Digunakan jika hasil positif (1) lebih signifikan (misal: hasil tes penyakit). Nilai $t$ diabaikan.
$$d(i, j) = \frac{r + s}{q + r + s}$$
*Persamaan kemiripannya disebut **Jaccard Coefficient**:*
$$sim(i, j) = \frac{q}{q + r + s}$$

### 3. Jika hanya satu kolom yang digunakan, maka jarak biner dapat dihitung dengan:
$$d(i, j) = \begin{cases} 0, & \text{jika } i = j \\ 1, & \text{jika } i \neq j \end{cases}$$

---

## Mengukur Jarak Tipe Categorical (Nominal)
Digunakan untuk atribut yang tidak memiliki urutan (misal: warna, profesi).

### 1. Overlay Metric (OM)
Ukuran paling sederhana, memberikan nilai 0 jika sama dan 1 jika berbeda.
$$d(x, y) = \sum_{i=1}^{n} \delta(a_i(x), a_i(y))$$
Di mana $\delta(a_i(x), a_i(y)) = 0$ jika nilai atribut sama, dan $1$ jika berbeda.

### 2. Value Difference Metric (VDM)
Mengasumsikan dua nilai atribut lebih dekat jika memiliki klasifikasi kelas ($C$) yang sama.
$$d(x, y) = \sum_{i=1}^{n} \sum_{c=1}^{C} |P(c|a_i(x)) - P(c|a_i(y))|$$

---

## Mengukur Jarak Tipe Ordinal
Atribut ordinal memiliki urutan (misal: kecil, sedang, besar). Prosedur penghitungannya adalah:
1. Ganti nilai atribut dengan peringkatnya ($r_{if} \in \{1, \dots, M_f\}$).
2. Lakukan normalisasi nilai ke rentang $[0, 1]$ agar memiliki bobot yang sama dengan tipe data lain:
$$z_{if} = \frac{r_{if} - 1}{M_f - 1}$$
3. Hitung jarak menggunakan ukuran numerik seperti **Euclidean** atau **Manhattan** pada nilai $z_{if}$ tersebut.

---

## Mengukur Jarak Tipe Campuran (Mixed Types)
Untuk database yang memiliki berbagai tipe data (numerik, biner, nominal, ordinal), digunakan **Gower’s Similarity Coefficient**:

$$d(i, j) = \frac{\sum_{f=1}^{p} \delta_{ij}^{(f)} d_{ij}^{(f)}}{\sum_{f=1}^{p} \delta_{ij}^{(f)}}$$

Keterangan:
- $\delta_{ij}^{(f)}$ adalah indikator (bernilai 0 jika data pada atribut $f$ hilang, atau jika atribut bersifat asimetris biner dan keduanya bernilai 0; selain itu bernilai 1).
- $d_{ij}^{(f)}$ adalah kontribusi jarak atribut ke-$f$:
  - Jika **Numerik/Ordinal**: $d_{ij}^{(f)} = \frac{|x_{if} - x_{jf}|}{R_f}$ (di mana $R_f$ adalah rentang nilai maksimum-minimum).
  - Jika **Biner/Nominal**: $d_{ij}^{(f)} = 0$ jika $x_{if} = x_{jf}$, dan $1$ jika sebaliknya.

# Implementasi

## Dataset Iris (Numerik)

In [ ]:
import pandas as pd
import numpy as np


df = pd.read_csv("dataset/IRIS.csv")
df.head()

,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,3.5,1.4,0.2,Iris-setosa
1,4.9,3.0,1.4,0.2,Iris-setosa
2,4.7,3.2,1.3,0.2,Iris-setosa
3,4.6,3.1,1.5,0.2,Iris-setosa
4,5.0,3.6,1.4,0.2,Iris-setosa


Tabel Tipe Data dan Penjelasan Kolom Dataset Iris

| Nama Kolom | Tipe Data (Pandas) | Jenis Atribut (Statistik) | Penjelasan Fungsi Kolom |
| :--- | :--- | :--- | :--- |
| **`sepal_length`** | `float64` | Numerik (Kontinu) | Ukuran panjang kelopak bunga (sepal) dalam satuan centimeter. |
| **`sepal_width`** | `float64` | Numerik (Kontinu) | Ukuran lebar kelopak bunga (sepal) dalam satuan centimeter. |
| **`petal_length`** | `float64` | Numerik (Kontinu) | Ukuran panjang mahkota bunga (petal) dalam satuan centimeter. |
| **`petal_width`** | `float64` | Numerik (Kontinu) | Ukuran lebar mahkota bunga (petal) dalam satuan centimeter. |
| **`species`** | `object` (String) | Kategorikal (Nominal) | Label kelas (target) yang berisi teks nama spesies bunga Iris (`Iris-setosa`, `Iris-versicolor`, `Iris-virginica`). |

Ringkasan hasil deskripsi data:



| Kategori Informasi | Jumlah | Nama Elemen | Deskripsi Singkat |
|---|---|---|---|
| **Dimensi Dataset** | 150 Baris, 5 Kolom | `sepal_length`, `sepal_width`, `petal_length`, `petal_width`, `species` | Total observasi sampel bunga Iris beserta seluruh variabelnya. |
| **Atribut** | 4 Atribut | `sepal_length`, `sepal_width`, `petal_length`, `petal_width` |
| **Label** | 3 Kelas | `species` (`Iris-setosa`, `Iris-versicolor`, `Iris-virginica`) | Variabel dependen yang menjadi tujuan akhir dari klasifikasi. |

## Perhitungan Manual Jarak Dataset Iris

Berdasarkan data sampel Iris (Indeks 0 dan Indeks 1):
- **P1 (Indeks 0):** (5.1, 3.5, 1.4, 0.2)
- **P2 (Indeks 1):** (4.9, 3.0, 1.4, 0.2)

---

### Euclidean Distance

Persamaan Umum:

$$d_{Euclidean}(p, q) = \sqrt{\sum_{i=1}^{n} (p_i - q_i)^2}$$

Implementasi pada fitur Iris:

$$d = \sqrt{(SL_p - SL_q)^2 + (SW_p - SW_q)^2 + (PL_p - PL_q)^2 + (PW_p - PW_q)^2}$$

Contoh Perhitungan:

$$d = \sqrt{(5.1 - 4.9)^2 + (3.5 - 3.0)^2 + (1.4 - 1.4)^2 + (0.2 - 0.2)^2}$$

$$d = \sqrt{(0.2)^2 + (0.5)^2 + 0^2 + 0^2}$$

$$d = \sqrt{0.04 + 0.25}$$

$$d = \sqrt{0.29} \approx 0.5385$$

---

### Manhattan Distance

Persamaan Umum:

$$d_{Manhattan}(p, q) = \sum_{i=1}^{n} |p_i - q_i|$$

Implementasi pada fitur Iris:

$$d = |SL_p - SL_q| + |SW_p - SW_q| + |PL_p - PL_q| + |PW_p - PW_q|$$

Contoh Perhitungan:

$$d = |5.1 - 4.9| + |3.5 - 3.0| + |1.4 - 1.4| + |0.2 - 0.2|$$

$$d = |0.2| + |0.5| + 0 + 0$$

$$d = 0.2 + 0.5$$

$$d = 0.7000$$


---

### Ringkasan Hasil

| Metode Perhitungan | Hasil Jarak |
| :--- | :--- |
| **Euclidean Distance** | **0.5385** |
| **Manhattan Distance** | **0.7000** |

### Euclidean Distance

In [27]:

cols = ['sepal_length', 'sepal_width', 'petal_length', 'petal_width']

p1 = df.loc[0, cols]
p2 = df.loc[1, cols]

euclidean = np.sqrt(np.sum(((p1 - p2))**2))

print(f"Hasil Jarak Euclidean : {euclidean:.4f}")

Hasil Jarak Euclidean : 0.5385


### Manhattan Distance

In [28]:
cols = ['sepal_length', 'sepal_width', 'petal_length', 'petal_width']

p1 = df.loc[0, cols]
p2 = df.loc[1, cols]

manhattan = np.sum(np.abs(p1 - p2))

print(f"Hasil Jarak Euclidean : {manhattan:.4f}")

Hasil Jarak Euclidean : 0.7000


## Dataset Titanic (Campuran)

In [5]:
import pandas as pd
import numpy as np


df = pd.read_csv("dataset/Titanic-Dataset.csv")
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [6]:
print(f"Dimensi dataset Iris: {df.shape[0]} baris dan {df.shape[1]} kolom.\n")

print("Daftar kolom/fitur dalam dataset:")
print(df.columns.tolist())

print("-" * 50)

print("\nInformasi detail tipe data:")
df.info()

Dimensi dataset Iris: 891 baris dan 12 kolom.

Daftar kolom/fitur dalam dataset:
['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']
--------------------------------------------------

Informasi detail tipe data:
<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
mem

### Tabel Tipe Data dan Penjelasan Kolom / Atribut

| Nama Kolom | Tipe Data (Pandas) | Jenis Atribut (Statistik) | Penjelasan Fungsi Kolom |
| :--- | :--- | :--- | :--- |
| **`PassengerId`** | `int64` | Numerik| ID unik untuk setiap penumpang. |
| **`Survived`** | `int64` | Binary | Status keselamatan (0 = Meninggal, 1 = Selamat). |
| **`Pclass`** | `int64` | Ordinal | Tingkatan kelas tiket (1 = Kelas 1, 2 = Kelas 2, 3 = Kelas 3). |
| **`Name`** | `str` (String) | Categorical | Nama lengkap penumpang. |
| **`Sex`** | `str` (String) | Binary| Jenis kelamin penumpang (`male` atau `female`). |
| **`Age`** | `float64` | Numerik| Usia penumpang dalam satuan tahun. |
| **`SibSp`** | `int64` | Numerik | Jumlah saudara/pasangan yang ikut di kapal. |
| **`Parch`** | `int64` | Numerik| Jumlah orang tua/anak yang ikut di kapal. |
| **`Ticket`** | `str` (String) | Categorical | Nomor tiket perjalanan. |
| **`Fare`** | `float64` | Numerik | Harga tiket atau tarif perjalanan. |
| **`Cabin`** | `str` (String) | Categorical | Nomor kabin penumpang (banyak data kosong). |
| **`Embarked`** | `str` (String) | Categorical | Pelabuhan keberangkatan (C, Q, S). |

---
---

### Kolom yang dipakai untuk perhitungan jarak

| Kategori Informasi | Jumlah | Nama Elemen | Deskripsi Singkat |
| :--- | :--- | :--- | :--- |
| **Atribut Numerik** | 4 Kolom | `Age`, `Fare`, `SibSp`, `Parch` | Digunakan untuk perhitungan jarak berbasis angka (Euclidean/Manhattan). |
| **Atribut Categorikal** | 1 Kolom | `Embarked` | Variabel kualitatif yang dihitung berdasarkan kesamaan (Matching). |
| **Atribut Binary** | 1 Kolom | `Sex` | Variabel yang memiliki dua nilai (misalnya, `male` atau `female`). |
| **Atribut Ordinal** | 1 Kolom | `Pclass` | Variabel kategori yang memiliki urutan/tingkatan kelas. |
| **Total Sampel** | 891 Baris | Data Penumpang | Seluruh observasi yang terdapat dalam dataset `Titanic-Dataset.csv`. |

---

### Data Sampel untuk Simulasi Perhitungan

| Indeks (ID) | Age | Fare | SibSp | Parch | Sex | Embarked | Pclass |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| **0 (P0)** | 22.0 | 7.25 | 1 | 0 | male | S | 3 |
| **1 (P1)** | 38.0 | 71.2833 | 1 | 0 | female | C | 1 |

---

## Numerik

Pada perhitungan numerik ini, saya menggunakan data kolom `Age`, `Fare`, `SibSp`, dan `Parch` untuk menghitung jarak antara dua penumpang (P0 dan P1) menggunakan beberapa ukuran jarak yang telah dijelaskan sebelumnya.

### Euclidean Distance (Normalized)

Jarak Euclidean adalah panjang garis lurus antara dua titik dalam ruang multidimensi. Rumus umumnya adalah:

Jika diterapkan pada fitur numerik Titanic dengan normalisasi berbasis Range ($R_k$):

$$d_{Euclidean\_Norm} = \sqrt{\left(\frac{Age_p - Age_q}{R_{Age}}\right)^2 + \left(\frac{Fare_p - Fare_q}{R_{Fare}}\right)^2 + \left(\frac{SibSp_p - SibSp_q}{R_{SibSp}}\right)^2 + \left(\frac{Parch_p - Parch_q}{R_{Parch}}\right)^2}$$

**Parameter Range ($R_k$):**

- $R_{Age} = 79.58$
- $R_{Fare} = 512.33$
- $R_{SibSp} = 8$
- $R_{Parch} = 6$

---

Substitusi nilai ke dalam rumus:

$$d = \sqrt{\left(\frac{22 - 38}{79.58}\right)^2 + \left(\frac{7.25 - 71.2833}{512.33}\right)^2 + \left(\frac{1 - 1}{8}\right)^2 + \left(\frac{0 - 0}{6}\right)^2}$$

$$d = \sqrt{\left(\frac{-16}{79.58}\right)^2 + \left(\frac{-64.0333}{512.33}\right)^2 + 0^2 + 0^2}$$

$$d = \sqrt{(-0.2011)^2 + (-0.1250)^2}$$

$$d = \sqrt{0.0404 + 0.0156}$$

$$d = \sqrt{0.0560} \approx 0.2366$$

In [15]:
cols = ['Age', 'Fare', 'SibSp', 'Parch']
ranges = df[cols].max() - df[cols].min()

p1 = df.loc[0, cols]
p2 = df.loc[1, cols]

euclidean = np.sqrt(np.sum(((p1 - p2) / ranges)**2))

print(f"Jarak Euclidean (Normalized) : {euclidean:.4f}")



Jarak Euclidean (Normalized) : 0.2367


### Perhitungan Manhattan Distance (Normalized)

Jika diterapkan pada fitur numerik Titanic dengan normalisasi berbasis Range ($R_k$):

$$d_{Manhattan\_Norm} = \sum_{k=1}^{n} \frac{|p_k - q_k|}{R_k}$$

$$d_{Manhattan\_Norm} = \frac{|Age_p - Age_q|}{R_{Age}} + \frac{|Fare_p - Fare_q|}{R_{Fare}} + \frac{|SibSp_p - SibSp_q|}{R_{SibSp}} + \frac{|Parch_p - Parch_q|}{R_{Parch}}$$

**Parameter Range ($R_k$):**

- $R_{Age} = 79.58$
- $R_{Fare} = 512.33$
- $R_{SibSp} = 8$
- $R_{Parch} = 6$

---

Substitusi nilai ke dalam rumus:

$$d = \frac{|22 - 38|}{79.58} + \frac{|7.25 - 71.2833|}{512.33} + \frac{|1 - 1|}{8} + \frac{|0 - 0|}{6}$$

$$d = \frac{16}{79.58} + \frac{64.0333}{512.33} + 0 + 0$$

$$d = 0.2011 + 0.1250 + 0 + 0$$

$$d = 0.3261$$

In [16]:
manhattan = np.sum(np.abs(p1 - p2) / ranges)
print(f"Jarak Manhattan (Normalized) : {manhattan:.4f}")

Jarak Manhattan (Normalized) : 0.3260


## Binary & Categorical

Pada perhitungan jarak untuk tipe data binary dan categorical, saya menggunakan data kolom `Sex` (Binary) dan `Embarked` (Categorical) untuk menghitung jarak antara dua penumpang (P0 dan P1) menggunakan ukuran jarak yang telah dijelaskan sebelumnya.

Pada tipe data ini, tidak ada urutan atau besaran angka. Jarak hanya dihitung berdasarkan apakah nilai tersebut identik atau tidak.

Jarak antara dua data $p$ dan $q$ untuk fitur kategorikal/biner didefinisikan sebagai:

$$d(p, q) = \begin{cases} 0, & \text{jika } p = q \\ 1, & \text{jika } p \neq q \end{cases}$$

Berdasarkan data sampel:
- **P0:** `Sex` = male, `Embarked` = S
- **P1:** `Sex` = female, `Embarked` = C

---

#### 1. Fitur Binary (`Sex`)
Perbandingan nilai untuk fitur jenis kelamin antara Penumpang 0 dan Penumpang 1:
- $p_{Sex} = \text{male}$
- $q_{Sex} = \text{female}$

Karena $p \neq q$, maka:
$$d_{Sex}(p, q) = 1$$

#### 2. Fitur Categorical (`Embarked`)
Perbandingan nilai untuk fitur pelabuhan keberangkatan antara Penumpang 0 dan Penumpang 1:
- $p_{Embarked} = \text{S}$
- $q_{Embarked} = \text{C}$

Karena $p \neq q$, maka:
$$d_{Embarked}(p, q) = 1$$

---

In [18]:
col_bin = 'Sex'
col_cat = 'Embarked'

p1_sex = df.loc[0, col_bin]
p2_sex = df.loc[1, col_bin]

p1_emb = df.loc[0, col_cat]
p2_emb = df.loc[1, col_cat]

dist_sex = 0 if p1_sex == p2_sex else 1
dist_emb = 0 if p1_emb == p2_emb else 1

print(f"Jarak Binary: {dist_sex}")
print(f"Jarak Categorical: {dist_emb}")

Jarak Binary: 1
Jarak Categorical: 1


## Ordinal

Pada perhitungan jarak untuk tipe data ordinal, saya menggunakan data kolom `Pclass` untuk menghitung jarak antara dua penumpang (P0 dan P1) menggunakan ukuran jarak yang telah dijelaskan sebelumnya.

### Perhitungan Jarak Ordinal (Pclass)

Pada fitur `Pclass` (Passenger Class), terdapat tingkatan kelas (1, 2, dan 3). Kita akan mengikuti prosedur normalisasi peringkat agar nilainya berada dalam rentang $[0, 1]$.

#### 1. Menentukan Peringkat ($r_{if}$) dan Nilai Maksimum ($M_f$)
Berdasarkan dataset, nilai kelas adalah 1, 2, dan 3. Maka nilai peringkat tertinggi ($M_f$) adalah **3**.
- Peringkat Penumpang 0 ($r_{0, Pclass}$) = **3**
- Peringkat Penumpang 1 ($r_{1, Pclass}$) = **1**

#### 2. Normalisasi Nilai ($z_{if}$)
Menggunakan rumus: 

$z_{if} = \frac{r_{if} - 1}{M_f - 1}$

- **Penumpang 0 (P0):**

  $$z_{0} = \frac{3 - 1}{3 - 1} = \frac{2}{2} = 1.0$$

- **Penumpang 1 (P1):**

  $$z_{1} = \frac{1 - 1}{3 - 1} = \frac{0}{2} = 0.0$$

#### 3. Menghitung Jarak Manhattan pada Nilai $z_{if}$
Jarak ordinal antara P0 dan P1 dihitung sebagai selisih absolut dari nilai yang telah dinormalisasi:

$$d_{Ordinal}(p, q) = |z_0 - z_1|$$

$$d_{Ordinal} = |1.0 - 0.0| = 1.0$$

---

In [19]:
col_ord = 'Pclass'

p1_class = df.loc[0, col_ord]
p2_class = df.loc[1, col_ord]

max_val = df[col_ord].max()
min_val = df[col_ord].min()

dist_pclass = abs(p1_class - p2_class) / (max_val - min_val)

print(f"Jarak Ordinal: {dist_pclass}")

Jarak Ordinal: 1.0


## Gower’s Similarity Coefficient (Tipe Campuran)

Berdasarkan rumus Gower, kita akan menjumlahkan seluruh kontribusi jarak dari setiap fitur ($d_{ij}^{(f)}$) lalu membaginya dengan jumlah fitur ($p$).

#### 1. Rekapitulasi Kontribusi Jarak ($d_{ij}^{(f)}$)

| Fitur ($f$) | Tipe | Nilai P0 | Nilai P1 | Perhitungan | Skor ($d^{(f)}$) |
| :--- | :--- | :--- | :--- | :--- | :--- |
| **Age** | Numerik | 22.0 | 38.0 | $|22-38| / 79.58$ | 0.2011 |
| **Fare** | Numerik | 7.25 | 71.28 | $|7.25-71.28| / 512.33$ | 0.1250 |
| **SibSp** | Numerik | 1 | 1 | $|1-1| / 8$ | 0.0000 |
| **Parch** | Numerik | 0 | 0 | $|0-0| / 6$ | 0.0000 |
| **Sex** | Biner | male | female | $male \neq female$ | 1.0000 |
| **Embarked** | Nominal | S | C | $S \neq C$ | 1.0000 |
| **Pclass** | Ordinal | 3 | 1 | $|1.0 - 0.0|$ (Norm) | 1.0000 |

---

#### 2. Substitusi ke Rumus Gower

Sesuai rumus:

$$d(i, j) = \frac{\sum_{f=1}^{p} \delta_{ij}^{(f)} d_{ij}^{(f)}}{\sum_{f=1}^{p} \delta_{ij}^{(f)}}$$

Karena semua data tersedia ($\delta_{ij} = 1$) dan total fitur ($p$) adalah **7**, maka:

$$d(0, 1) = \frac{0.2011 + 0.1250 + 0 + 0 + 1 + 1 + 1}{7}$$

$$d(0, 1) = \frac{3.3261}{7}$$

$$d(0, 1) \approx 0.4751$$

---

In [20]:
list_atribut = ["Age", "Fare", "SibSp", "Parch", "Sex", "Embarked", "Pclass"]
list_jarak = [manhattan, dist_sex, dist_emb, dist_pclass]
hasil_gower = sum(list_jarak) / len(list_atribut)

print(f"Total Jarak Gower (P0 vs P1): {hasil_gower:.4f}")

Total Jarak Gower (P0 vs P1): 0.4751


### Kesimpulan Akhir
Hasil akhir perhitungan **Gower Distance** antara Penumpang 0 dan Penumpang 1 adalah **0.4751**. 

Nilai ini merepresentasikan rata-rata ketidaksamaan dari seluruh tipe atribut. Semakin mendekati **0**, maka kedua objek semakin serupa. Dalam kasus ini, skor **0.4751** menunjukkan tingkat perbedaan yang moderat (sekitar 47.51%) yang dipicu oleh perbedaan signifikan pada kelas tiket, jenis kelamin, dan pelabuhan asal, meskipun memiliki kemiripan pada jumlah anggota keluarga di kapal.